File: CoinTrader/utils/IndicatorMaker.swift
//
//  IndicatorMaker.swift
//  CoinTrader
//
//  Created by choi hyunjin on 2022/02/05.
//

import Foundation

class IndicatorMaker {
	let n: Int
	let k: Int
	let priceManager = PricesManager()
	var ma = 0.0
	
	var touchedBottom = false
	var touchedTop = false
	
	
	init(n: Int, k: Int){
		self.n = n
		self.k = k
	}
	
	convenience init(){
		self.init(n: 20, k:2)
	}
	
	func setPrices(_ prices:[Double]){
		priceManager.prices = prices
	}
	
	func lower() -> Double{
		return self.ma - self.priceManager.getStandardDeviation(during: self.n) * Double(self.k)
	}
	func upper() -> Double{
		return self.ma + self.priceManager.getStandardDeviation(during: self.n) * Double(self.k)
	}
	
	func movingAverage(during: Int) -> Double{
		return self.priceManager.getAverage(during: during)
	}
	
	func pushPrice(price: Double){
		self.priceManager.pushPrice(price: price)
		self.ma = Double(self.movingAverage(during: self.n))
		
		if self.lower() >= price {
			self.touchedBottom = true
		}else if self.upper() <= price {
			self.touchedTop = true
		}
	}
	
	func getBW() -> Double {
		return (self.upper() - self.lower()) / self.ma
	}
	
	func resetTouched () {
		self.touchedTop = false
		self.touchedBottom = false
	}
}

File: src/engine/IndicatorMaker.ts
import {PricesManager} from './PricesManager';

export class IndicatorMaker {
  readonly n: number;
  readonly k: number;
  readonly pricesManager: PricesManager;
  ma: number = 0;
  touchedBottom: boolean = false;
  touchedTop: boolean = false;

  constructor(n: number = 20, k: number = 2) {
    this.n = n;
    this.k = k;
    this.pricesManager = new PricesManager(n);
  }

  setPrices(prices: number[]): void {
    this.pricesManager.setPrices(prices);
    if (this.pricesManager.prices.length >= this.n) {
      this.ma = this.movingAverage(this.n);
    }
  }

  lower(): number {
    return this.ma - this.pricesManager.getStandardDeviation(this.n) * this.k;
  }

  upper(): number {
    return this.ma + this.pricesManager.getStandardDeviation(this.n) * this.k;
  }

  movingAverage(period: number): number {
    return this.pricesManager.getAverage(period);
  }

  pushPrice(price: number): void {
    this.pricesManager.pushPrice(price);
    this.ma = this.movingAverage(this.n);

    if (this.lower() >= price) {
      this.touchedBottom = true;
    } else if (this.upper() <= price) {
      this.touchedTop = true;
    }
  }

  getBW(): number {
    if (this.ma === 0) {
      return 0;
    }
    return (this.upper() - this.lower()) / this.ma;
  }

  resetTouched(): void {
    this.touchedTop = false;
    this.touchedBottom = false;
  }
}
